templates/boltzmann/core/prelude.ts

```diff
@@ -28,31 +28,254 @@ function _getServiceName() {
+import { RedisInstrumentation } from '@opentelemetry/instrumentation-redis'
+void `{% endif %}`
+
+void `{% if postgres %}`
+import { PgInstrumentation } from '@opentelemetry/instrumentation-pg'
+void `{% endif %}`
+
+function normalizeHoneycombEnvVars (env: typeof process.env): void {
+  if (!env.HONEYCOMB_DATASET && env.HONEYCOMBIO_DATASET) {
+    env.HONEYCOMB_DATASET = env.HONEYCOMBIO_DATASET
+  }
 
-if (!process.env.HONEYCOMB_DATASET && process.env.HONEYCOMBIO_DATASET) {
-  process.env.HONEYCOMB_DATASET = process.env.HONEYCOMBIO_DATASET
+  if (!env.HONEYCOMB_WRITEKEY && env.HONEYCOMBIO_WRITEKEY) {
+    env.HONEYCOMB_WRITEKEY = process.env.HONEYCOMBIO_WRITEKEY
+  }
+
+  if (!env.HONEYCOMB_SAMPLE_RATE && env.HONEYCOMBIO_SAMPLE_RATE) {
+    env.HONEYCOMB_SAMPLE_RATE = env.HONEYCOMBIO_SAMPLE_RATE
+  }
+
+  if (!env.HONEYCOMB_TEAM && env.HONEYCOMBIO_TEAM) {
+    env.HONEYCOMB_TEAM = env.HONEYCOMBIO_TEAM
+  }
+
+  if (!env.HONEYCOMB_DATASET && env.HONEYCOMBIO_DATASET) {
+    env.HONEYCOMB_DATASET = env.HONEYCOMBIO_DATASET
+  }
 }
 
-if (!process.env.HONEYCOMB_WRITEKEY && process.env.HONEYCOMBIO_WRITEKEY) {
-  process.env.HONEYCOMB_WRITEKEY = process.env.HONEYCOMBIO_WRITEKEY
+function isHoneycomb (env: typeof process.env): boolean {
+  return !!env.HONEYCOMB_WRITEKEY
 }
 
-if (!process.env.HONEYCOMB_SAMPLE_RATE && process.env.HONEYCOMBIO_SAMPLE_RATE) {
-  process.env.HONEYCOMB_SAMPLE_RATE = process.env.HONEYCOMBIO_SAMPLE_RATE
+function isOtel (env: typeof process.env): boolean {
+  if (isHoneycomb(env) && env.HONEYCOMB_API_HOST) {
+    return /^grpc:\/\//.test(env.HONEYCOMB_API_HOST)
+  }
+  return false
 }
 
-if (!process.env.HONEYCOMB_TEAM && process.env.HONEYCOMBIO_TEAM) {
-  process.env.HONEYCOMB_TEAM = process.env.HONEYCOMBIO_TEAM
+function getHoneycombSampleRate (env: typeof process.env): number {
+  const rate = Number(env.HONEYCOMB_SAMPLE_RATE || 1)
+  assert(!isNaN(rate))
+  return rate
 }
 
-beeline({
-  writeKey: process.env.HONEYCOMB_WRITEKEY,
-  dataset: process.env.HONEYCOMB_DATASET,
-  sampleRate: Number(process.env.HONEYCOMB_SAMPLE_RATE) || 1,
-  serviceName,
-})
+function getHoneycombApiHost (env: typeof process.env): string {
+  assert(typeof env.HONEYCOMB_API_HOST === 'string')
+  return env.HONEYCOMB_API_HOST as string
+}
+
+function createOtelMetadata (env: typeof process.env): Metadata {
+  const metadata = new Metadata()
+
+  assert([
+    env.HONEYCOMB_WRITE_KEY,
+    env.HONEYCOMB_DATASET
+  ].every(v => typeof v === 'string' && v.length))
+
+  metadata.set('x-honeycomb-team', <string>env.HONEYCOMB_WRITE_KEY)
+  metadata.set('x-honeycomb-dataset', <string>env.HONEYCOMB_DATASET)
+  return metadata
+}
+
+function createOtelSampler (sampleRate: number): Sampler {
+  let sampler: Sampler = new AlwaysOnSampler()
+
+  if (sampleRate === 0) {
+    sampler = new AlwaysOffSampler()
+  } else if (sampleRate !== 1) {
+    sampler = new ParentBasedSampler({
+      root: new TraceIdRatioBasedSampler(sampleRate)
+    })
```

Comment:
I think this logic might be slightly off.

Even when the sampleRate is 1, we still want to use a ParentBasedSampler--because even if the current service is set to not sample, the service that's calling it might be set to sample! And we want to inherit its sampling decision :)

Reply:
I have no idea if it's worth in terms of efficiency it to special-case 0 and 1, or just do the simplest thing of always using this 🤷 
```
sampler = new ParentBasedSampler({
      root: new TraceIdRatioBasedSampler(sampleRate)
```

Reply:
Yeah, you know what? You're right. There's no particularly good reason for us to even do this optimization and it's definitely not worth us having to ask these kinds of questions. I'll drop that functionality and call it a day.